WordPress/gutenberg
# Mobile - Notice list - Code improvements and bug fixes

- `Gutenberg Mobile PR` -> https://github.com/wordpress-mobile/gutenberg-mobile/pull/4768

Fixes https://github.com/wordpress-mobile/WordPress-iOS/issues/18036
Fixes https://github.com/wordpress-mobile/WordPress-Android/issues/15643

## What?
A few issues were opened reporting that toggling between the HTML / Visual mode of the editor does not track the change in the notices.

## Why?
By checking the code I found the notices weren't getting removed from the store as well as the animation when collapsing wasn't running as expected.

## How?

For the Notice component, it refactors the code with the following:
- Replaces the Dimensions listener with `useWindowDimensions`.
- Handles the start of the animation and the hiding callback separately, this fixes calling `onNoticeHidden` when it should, to remove the displayed notice from the store.

For the NoticeList component:
- Refactors the component into a functional one.
- It removes the `shouldStack` prop since it's not used at all.
- Removes the `key` prop from the containers `View`.

## Testing Instructions

### Test case 1 - Toggling between HTML/Visual mode

- Open the editor
- Toggle to HTML mode
- Quickly Toggle to Visual mode
- Again Toggle to HTML mode
- **Expect** to see the notices to be synced with the current editor mode

### Test case 2 - Block actions

- Open the editor
- Type some text
- Using the block actions try copying, duplicating a block
- **Expect** to see the notice messages show correctly

## Screenshots or screencast 

Test case 1 (before) | Test case 1 (after)
-|-
<kbd><img src="https://user-images.githubusercontent.com/4885740/163819053-30770027-de02-476d-b0da-1717e55669c6.gif" width="250" /></kbd> | <kbd><img src="https://user-images.githubusercontent.com/4885740/163819064-e140128a-20a7-4c08-af53-b1d680160bbd.gif" width="250" /></kbd>

Test case 2 (before) | Test case 2 (after)
-|-
<kbd><img src="https://user-images.githubusercontent.com/4885740/163819120-673824e0-a2fa-4278-b351-e179dd3db394.gif" width="250" /></kbd> | <kbd><img src="https://user-images.githubusercontent.com/4885740/163819131-92b84e35-1b26-4425-bd18-8c26c8db735c.gif" width="250" /></kbd>
